safehttp/header.go

```diff
@@ -0,0 +1,134 @@
+
+// Header represents the key-value pairs in an HTTP header.
+// The keys will be in canonical form, as returned by
+// textproto.CanonicalMIMEHeaderKey.
+type Header struct {
+	wrappedHeader http.Header
+	immutable     map[string]bool
+}
+
+func newHeader(h http.Header) Header {
+	return Header{wrappedHeader: h, immutable: map[string]bool{}}
+}
+
+// MarkImmutable marks the header with the given name as immutable.
+// This header is now read-only. The name is first canonicalized
+// using textproto.CanonicalMIMEHeaderKey.
+func (h Header) MarkImmutable(name string) {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return
+	}
+	h.immutable[name] = true
+}
+
+// Set sets the header with the given name to the given value.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If this headers is not immutable, this function removes all other
+// values currently associated with this header before setting the new
+// value. Returns an error when applied on immutable headers.
+func (h Header) Set(name, value string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Set(name, value)
+	return nil
+}
+
+// Add adds a new header with the given name and the given value to
+// the collection of headers. The name is first canonicalized using
+// textproto.CanonicalMIMEHeaderKey. Returns an error when applied
+// on immutable headers.
+func (h Header) Add(name, value string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Add(name, value)
+	return nil
+}
+
+// Del deletes all headers with the given name. The name is first
+// canonicalized using textproto.CanonicalMIMEHeaderKey. Returns an
+// error when applied on immutable headers.
+func (h Header) Del(name string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Del(name)
+	return nil
+}
+
+// Get returns the value of the first header with the given name.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If no header exists with the given name then "" is returned.
+func (h Header) Get(name string) string {
+	return h.wrappedHeader.Get(name)
+}
+
+// Values returns all the values of all the headers with the given name.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If no header exists with the name `name` then nil is returned.
+func (h Header) Values(name string) []string {
+	return h.wrappedHeader.Values(name)
```

Comment:
interesting that if we don't do a "clone" here the caller will be able to replace some values, even for the immutable headers, by replacing entries in the slice.

@kele do we want to do smth about this?

Reply:
Ah, bugs! I think we definitely should make a clone

Reply:
Added cloning in 9692b10